security/pkg/server/ca/authenticate/xfcc_authenticator.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authenticate
+
+import (
+	"fmt"
+	"net/http"
+
+	"google.golang.org/grpc/metadata"
+
+	"github.com/alecholmes/xfccparser"
+
+	"istio.io/istio/pkg/security"
+)
+
+const (
+	XfccAuthenticatorType = "XfccAuthenticator"
+)
+
+// XfccAuthenticator extracts identities from Xfcc header.
+type XfccAuthenticator struct{}
+
+var _ security.Authenticator = &XfccAuthenticator{}
+
+func (xff XfccAuthenticator) AuthenticatorType() string {
+	return XfccAuthenticatorType
+}
+
+// Authenticate extracts identities from Xfcc Header.
+func (xff XfccAuthenticator) Authenticate(ctx *security.AuthContext) (*security.Caller, error) {
+	// First check if client is trusted client so that we can "trust" the Xfcc Header.
+	delegated := &CidrAuthenticator{}
+	if !delegated.CanTrustCaller(ctx) {
+		return nil, fmt.Errorf("caller is not in the trusted network. XfccAuthenticator can not be used")
+	}
+	meta, ok := metadata.FromIncomingContext(ctx.RequestContext)
+
+	if !ok || len(meta.Get(xfccparser.ForwardedClientCertHeader)) == 0 {
+		return nil, nil
+	}
+	xfccHeader := meta.Get(xfccparser.ForwardedClientCertHeader)[0]
+	return buildSecurityCaller(xfccHeader)
+}
+
+// AuthenticateRequest validates Xfcc Header.
+func (xff XfccAuthenticator) AuthenticateRequest(req *http.Request) (*security.Caller, error) {
+	xfccHeader := req.Header.Get(xfccparser.ForwardedClientCertHeader)
+	if len(xfccHeader) == 0 {
+		return nil, nil
+	}
+	return buildSecurityCaller(xfccHeader)
+}
+
+func buildSecurityCaller(xfccHeader string) (*security.Caller, error) {
+	clientCerts, err := xfccparser.ParseXFCCHeader(xfccHeader)
+	if err != nil {
+		return nil, fmt.Errorf("error in parsing xfcc header: %v", err)
+	}
+	ids := []string{}
+	for _, cc := range clientCerts {
```

Comment:
Is this correct ? Does it return the LEAF cert only, or the chain ? 

If client presents multiple identities ( which I don't think Istio will ever do, and is not very easy ) we should select the URI starting with spiffee and having one of our trust domains. 
Or just reject the request - a client with multiple workload identities is not well defined and too early to support. Let's keep it simple...

Reply:
This is all the certs in the call chain - for which ever proxies it has passed through. Typically it will be one. I have added validations to reject if it zero or more than 1.

Reply:
The remote identity is in the first cert I believe - it is fine to have the identity of the other proxies, but mostly irrelevant, it's whatever the proxy in front of us happens to claim. We can't verify anything - just trust the one in front of us. May be useful for debugging/tracing/logging - but for auth it's not relevant I think, we make no use of the proxy path.

What we need to be sure is that identity of a proxy in front doesn't accidentally get treated as identity of the peer - proxies tend to have high SDS permissions. AFAIK Istiod makes all decisions based on the final namespace/KSA extracted from the initial mTLS ( first trusted proxy in the chain ) - I never fully understand why we pass all those identities around in the interface, but they should certainly not include any of the proxies in the chain.